### ISSUE TITLE
GC runs but working set remains the same?

### ISSUE THREAD

**Issue by Leonardo-Ferreira** [dotnet/runtime#90640]
Take a look here:
![image](https://github.com/dotnet/runtime/assets/5585339/cd1bc81f-f195-4ac6-8c96-426ced5e3e47)
after the GC ran It reported:
![image](https://github.com/dotnet/runtime/assets/5585339/da98a33a-1b89-4309-b949-b5f92fe5a0e5)

why didn't the working set dropped by aprox 19MB?

**Comment by mangod9** [comment]
The GC wouldnt immediately decommit memory unless there is memory pressure. That helps with overall performance since frequent commit/decommits arent required. 

The working set also depends on whether you are running with SVR, since with it enabled the GC uses as many heaps as the # of cores on the machine. In .NET 8 there is a new adaptive GC mode which can ramp up the number of heaps based on application load.

**Comment by teo-tsirpanis** [comment]
You can force the GC to decommit memory by calling `GC.Collect(GC.MaxGeneration, GCCollectionMode.Aggressive)`. I don't know if the [conserve memory](https://learn.microsoft.com/en-us/dotnet/core/runtime-config/garbage-collector#conserve-memory) options would have any effect here.

**Comment by Leonardo-Ferreira** [comment]
well, both comments are really interesting...  
first @mangod9 
> The GC wouldnt immediately decommit memory unless there is memory pressure

That's a bit concerning, you see, im trying to pack as many pods on Kubernetes as possible, and currently my biggest enemy is memory. My machines are already "memory optimized" and yet, cpu consumption is around 30-40% while memory is 85-90%... so, if I could, I would like to instruct the GC/SO to commit and decommit and not to worry about the CPU cost of it... I wonder how much of that 136 working set is REALLY required and how much "it's being kept around just in case"... I know for a fact now that at least 16MB there are "wasted", if that's true for all the pods (I have pods that are "locked" on 500+mb memory, but I won't take those into consideration), we're talking about a potential saving of more than $4k per month on that Kubernetes cluster alone because I would be able to decommission 2-3 nodes...

> The working set also depends on whether you are running with SVR, since with it enabled the GC uses as many heaps as the # of cores on the machine

what is SVR? and given that most pods there are allocated with a `requires` of "1m" and the `limit` of "150m", they should have only 1 heap...

now, @teo-tsirpanis,
I wouldn't like to run GC manually, but that conserve memory option is indeed very attractive, I will surely fiddle with it and see if I can get a part of the working set back...

**Comment by Maoni0** [comment]
it would be worse for perf to make the workingset shrink by approx 19mb, unless you never allocate again, ie, you'll never need to use that 19mb again. I explained the decommit behavior of the ephemeral segment [here](https://github.com/Maoni0/mem-doc/blob/master/doc/.NETMemoryPerformanceAnalysis.md#special-handling-of-the-ephemeral-segment). please take a look and let me know if that does not answer your question. 

now, of course we could discuss if this 19mb is too big for your scenario. this is your gen0 allocation budget, meaning when you have allocated 19mb in gen0, a GC will be triggered. since your other generations are so small, we could make this budget smaller and still achieve acceptable perf (what is your time-in-gc counter value?). as @mangod9 mentioned, this is exactly one of the scenarios we are targeting to make better with our 8.0 work. let me know if that's something you'd like to try out (I do want to make changes to that logic soon but will be happy to give you a private build of clrgc.dll/libclrgc when I'm done with that change).

**Comment by Leonardo-Ferreira** [comment]
@teo-tsirpanis setting `DOTNET_GCConserveMemory` to `9` did not changed anything basically... as far as I could observe the behavior was the same, both in terms of memory consumption and api response time...

hey @Maoni0 thanks for that article, ill check it out for sure, but do you wanna know something that made the memory drop to this level here?:
<img width="628" alt="image" src="https://github.com/dotnet/runtime/assets/5585339/601211ab-3092-4a7b-8d08-8eb88477bce5">

Memory Pressure!

I installed stress-ng and ran `stress-ng --brk 1 --stack 1 --bigheap 1 -t 5`

right out of the gate, according to `dotnet counters` the memory went back to 12 freaking mb! let me say it again, **TWELVE MEGABYTES**... right after it a ran a load using bombardier, and yes, the first few requests struggled but after that it was all life as usual in terms of response time. BUT in terms of memory usage, its now on a constant around 68-72... I even ran more aggressive loads but sure enough, it goes back down to its 70's...

for me this is kinda HUGE... the pod is used to launch with a working set of 100MB+...

**Comment by Leonardo-Ferreira** [comment]
@Maoni0 this API here is a really small one... we are talking less than 1 request per second at peak hours! on the middle of the night, this api could even be put to sleep! this api basically receives a request and "replays" it to two other apis and combine the results which is usually a 1kb or less json from each side... the "requires" in term of CPU here is `100m` but we never seen it go over 15%... the biggest surge we have ever observed was 12 requests at the same time... I mean... im not looking for performance in terms of "it runs fast", im looking for performance in terms that "it responds under a second using as little memory as possible"... and the thing is: I have about 50 others JUST LIKE THIS...

**Comment by Leonardo-Ferreira** [comment]
Looking at what can be done, I ended up testing upgrading this API to .Net 7 and to my surprise the overall performance dropped a bit and the working set raised to 195MB+... running the stress generator brings the memory down to around 125MB and after the rewarming the set the memory stays at around 165MB... I think I will keep it on .net 6 for now

**Comment by MichalPetryka** [comment]
> the overall performance dropped a bit and the working set raised to 195MB+

Try using `DOTNET_GCName=clrgc.dll`, it'll likely revert that regression.

**Comment by En3Tho** [comment]
@Leonardo-Ferreira It's still important to know whether your app runs in server (multi heap) or workstation (1 heap) mode. Heaps here is not something kubernetes provides, it's an internal memory space of GC to allocate stuff.

Also, when running in the kubernetes it might be worthy to set heap hard limit to exact value you've set in the pod limits.

My understanding is that Kubernetes is much more eager to kill a pod when memory  limit is reached while a typical os is more like "I can't give you more memory but please deal with it yourself"

**Comment by Leonardo-Ferreira** [comment]
@En3Tho thank you for the heads up, I will test setting the heap hard limit.

Regarding the server vs. workstation, how can I determine that? my project csproj file makes no reference to the `ServerGarbageCollection` and I don't specify any environment variable... given that the pod is capped at 256MB and everything else I've said above, what "flavor" should I be using?

**Comment by En3Tho** [comment]
I don't remember if there was a property on GC class or something but you can always fix the heap count via runtimeconfig.json or env variable. I guess you have a workstation mode on jusjing from your description. 

Which to choose is generally: 

Workstation works best when you have lots of small workers that share memory, or worker that is not high load/allocation pressure.

Server is more like "I'm the lord of these memory rich lands, it's all mine". E.g. webserver handling lots of requests and allocating and cleaning a lot.

In your case workstation mode seems like a fit.

**Comment by Leonardo-Ferreira** [comment]
Something else to make things complicated: I was working on a API that leaks memory to help me study and get better at reading dumps... I was upset that, in my point of view, `Marshal.AllcHCGlobal` was not leaking, because the subsequent `currentProcess.WorkingSet64` was not reporting the amount I was expecting and even worst, if I gave it enough time, it would basically return to the original value of the API... to my surprise, `WorkingSet64` was basically reporting RSS, and the 500MB that was actually leaked, happened to be paged out... lesson learned: low working set doesn't mean low memory usage...

**Comment by Leonardo-Ferreira** [comment]
> It's still important to know whether your app runs in server (multi heap) or workstation (1 heap) mode. Heaps here is not something kubernetes provides, it's an internal memory space of GC to allocate stuff.
> 
> Also, when running in the kubernetes it might be worthy to set heap hard limit to exact value you've set in the pod limits.
> 
> My understanding is that Kubernetes is much more eager to kill a pod when memory limit is reached while a typical os is more like "I can't give you more memory but please deal with it yourself"

I have been reading about [DATAS](https://maoni0.medium.com/dynamically-adapting-to-application-sizes-2d72fcb6f1ea) and it does seem very promissing...

**Comment by zhershds-notime** [comment]
@Leonardo-Ferreira , have you tried DATAS, did it help with your problem?

**Comment by Leonardo-Ferreira** [comment]
in this case here, not very much actually... the memory profile remains basically the same in .net 8 with or without DATAS... in other apis, that consume more memory and have more CPU/requests/work I saw better results, but in small apis, with very low traffic, no real gain

**Comment by typesbb** [comment]
> in this case here, not very much actually... the memory profile remains basically the same in .net 8 with or without DATAS... in other apis, that consume more memory and have more CPU/requests/work I saw better results, but in small apis, with very low traffic, no real gain...
> 
> I couldn't test yet with .Net 9... were there any improvements on DATAS or GC?

I repeat.

**Comment by Leonardo-Ferreira** [comment]
answering my own question, yes, it looks like .net 9 is bringing improvements to DATAS: https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-9/runtime#garbage-collection

**Comment by mangod9** [comment]
correct, DATAS has been enabled by default in 9. Note that it should help with most scenarios but since it's scenario dependent not all apps will observe gains. Also as might be expected there is always a memory tradeoff with throughput.

**Comment by janvorli** [comment]
@Leonardo-Ferreira going through backlog of GC issues, do you still have the same problems with recent .NET (>=10.0)?

**Comment by Leonardo-Ferreira** [comment]
I stepped away from that context @janvorli but I will ask the team and let you know

**Comment by janvorli** [comment]
Thank you!
